### ISSUE TITLE
Use PETSc's MatNullSpace to represent null space.

### ISSUE THREAD

**Issue by haplav** [It4innovations/permon#16]
Related to #13 .

**Comment by jkruzik** [comment]
I do not like the idea. MatNullSpace is just an array of vectors which does not have any basic Mat operations (e.g., MatMult). We are mostly using null space matrices to multiply vecs/mats.

**Comment by haplav** [comment]
OK, I was thinking about converting to a matrix if needed. But for the time being it's probably not worth the effort. There are more important things. Closing.

**Comment by haplav** [comment]
After all, I think it's still worth considering.
* We could have a routine which converts it to MATDENSE cheaply. And MatMat operations internally work with columns of R anyway.
* NullSpace would be naturally attached to the operator.
* The nullspace computation using MUMPS could be contributed to PETSc directly. The code would probably even become simpler. But then use of MatNullSpace would be required to fit into PETSc.

**Comment by haplav** [comment]
At the very least, we could start by attaching the nullspace to the operator directly using PetscObjectCompose. It shouldn't be a property of a QP nor MATINV.

**Comment by haplav** [comment]
> It shouldn't be a property of a QP nor MATINV.

See also #30.

**Comment by jkruzik** [comment]
> We could have a routine which converts it to MATDENSE cheaply.

When would you convert it? Would you store it? If so where? May I, suggest storing it in qp->R? ;)
You can have a sparse null space.
> And MatMat operations internally work with columns of R anyway.

Not when both matrices are of the same type e.g., AIJ or dense. Plain matvecs do not work with columns directly.
> The nullspace computation using MUMPS could be contributed to PETSc directly.

Sure, there is MatNullSpaceCreateRigidBody() so there could be something like MatNullSpaceCreateWithMumps().
> The code would probably even become simpler.

The code in PETSc yes but we would need a wrapper over such routine to handle our MatTypes (block diag).
> But then use of MatNullSpace would be required to fit into PETSc.

For the PETSc routine yes, but that does not mean we would need to use MatNullSpace in PERMON.

> At the very least, we could start by attaching the nullspace to the operator directly using PetscObjectCompose. It shouldn't be a property of a QP nor MATINV.

I do not see why it is better to have R attached to Mat operator instead of it being a pointer in the QP struct. However, I do see quite a few disadvantages:
- When and how would you destroy R if it is computed by PERMON?
- You would need PetscObjectQuery() in each function you wanted to use R which means a for loop with Strcmp().
- How would a user provide R? MatSetNullSpace works only with MatNullSpace struct. Do you want users to use PetscObjectCompose with a specific string?

To summarize:
1. Having nullspace as a matrix is convenient with our use case
2. QPSetOperator() followed by an optional QPSetOperatorNullSpace() provides a clean interface.

**Comment by haplav** [comment]
> When would you convert it? Would you store it? If so where? May I, suggest storing it in qp->R? ;)
You can have a sparse null space.

I though of an instantaneous conversion just once needed - e.g. right before MatMult. It could even be a Get/Restore pair.

> You can have a sparse null space.

I'm not sure sparse nullspace makes any sense. It's actually a _basis_ of the nullspace and hence a set of vectors even mathematically.

Nevertheless, I don't see a point in continuing this discussion about MatNullSpace now - it will pop up again once the nullspace computation gets to PETSC.

I will still continue to advocate moving R directly to the operator.
> When and how would you destroy R if it is computed by PERMON?

Composed objects are dereferenced/destroyed automatically once the parent object gets destroyed.

> You would need PetscObjectQuery() in each function you wanted to use R which means a for loop with Strcmp().
> How would a user provide R? MatSetNullSpace works only with MatNullSpace struct. Do you want users to use PetscObjectCompose with a specific string?

We can have a convenience functions `MatSetNullSpaceMatrix`/`MatGetNullSpaceMatrix` which use object composition internally. Users don't even need to know about this implementation detail.

> QPSetOperator() followed by an optional QPSetOperatorNullSpace() provides a clean interface.

I don't think it's cleaner than what I propose above. Additionally, my suggestion reflects the mathematical fact that the nullspace is a feature of the operator K. It's not at all a feature of the QP having K as Hessian, nor of K+.

I also think that my #30 reveals clearly it's not a good idea to have `MatInvSetNullSpace` and `QPSetNullSpace` in parallel. And [ex3.c](https://github.com/permon/permon/blob/haplav/generalize-MatInvComputeNullSpace/src/tests/ex3.c) in #23 shows it is awkward to need to create a MATINV object if I am just interested in the nullspace.

**Comment by jkruzik** [comment]
> > When would you convert it? Would you store it? If so where? May I, suggest storing it in qp->R? ;)
> > You can have a sparse null space.
> 
> I though of an instantaneous conversion just once needed - e.g. right before MatMult. It could even be a Get/Restore pair.
>
Am I understanding correctly that you would do a conversion every time before a MatMult? Do you realize that such a conversion needs to copy all the vectors values into a contiguous array?
> > You can have a sparse null space.
> 
> I'm not sure sparse nullspace makes any sense. It's actually a _basis_ of the nullspace and hence a set of vectors even mathematically.
> 
It does make sense. There is something called sparse null space problem which is related e.g., to optimization with linear equality constraints. A standard basis is clearly sparse. Of course, a basis is a set of vectors but so is a matrix.
> Nevertheless, I don't see a point in continuing this discussion about MatNullSpace now - it will pop up again once the nullspace computation gets to PETSC.
> 
> I will still continue to advocate moving R directly to the operator.
> 
> > When and how would you destroy R if it is computed by PERMON?
> 
> Composed objects are dereferenced/destroyed automatically once the parent object gets destroyed.
> 
You are right it is done when a header is destroyed.
> > You would need PetscObjectQuery() in each function you wanted to use R which means a for loop with Strcmp().
> > How would a user provide R? MatSetNullSpace works only with MatNullSpace struct. Do you want users to use PetscObjectCompose with a specific string?
> 
> We can have a convenience functions `MatSetNullSpaceMatrix`/`MatGetNullSpaceMatrix` which use object composition internally. Users don't even need to know about this implementation detail.
> 
Fair enough. What about having a couple of thousands of multiplications with R where each one does an unnecessary loop with strcmp?
> > QPSetOperator() followed by an optional QPSetOperatorNullSpace() provides a clean interface.
> 
> I don't think it's cleaner than what I propose above. Additionally, my suggestion reflects the mathematical fact that the nullspace is a feature of the operator K. It's not at all a feature of the QP having K as Hessian, nor of K+.
>
If you have MatSetNullSpaceMatrix than from the user standpoint it is the same as having QPSetOperatorNullSpace.

I wouldn't mind if it is part of the matrix. I would mind if there was a performance hit due to it not being a part of a readily accessible struct.

> I also think that my #30 reveals clearly it's not a good idea to have `MatInvSetNullSpace` and `QPSetNullSpace` in parallel.

I don't see what are you referring to. Can you be more specific?

> And [ex3.c](https://github.com/permon/permon/blob/haplav/generalize-MatInvComputeNullSpace/src/tests/ex3.c) [](/permon/permon/blob/HEAD@{2020-05-08T08:02:07Z}/generalize-MatInvComputeNullSpace/src/tests/ex3.c) in #23 shows it is awkward to need to create a MATINV object if I am just interested in the nullspace.

Well, then why won't we move it out of MatInv (could be even into PETSc)? One problem though, you need to be able to avoid multiple factorizations with the same matrix.

We could propose MatNullSpace having both vecs and mat, being able to set either and get one or other. The default GS for MatRemoveNullSpace() can do I-RR^T when the mat is set.

**Comment by haplav** [comment]
> Am I understanding correctly that you would do a conversion every time before a MatMult? Do you realize that such a conversion needs to copy all the vectors values into a contiguous array?

You can do that at a cost of single pointer copy in case of a dense matrix. Speaking about sparse matrices it indeed doesn't make much sense.

> It does make sense. There is something called sparse null space problem which is related e.g., to optimization with linear equality constraints. A standard basis is clearly sparse. Of course, a basis is a set of vectors but so is a matrix.

Fair enough. The sparse null space problem obviously exists. I was maybe too focused on the nullspace basis generated by MUMPS which makes no effort to sparsify it.

> I don't see what are you referring to. Can you be more specific?

I mean maintaining independent pointers in QP/MATINV and two separate setters/getters. I hate this redundancy.

>  Fair enough. What about having a couple of thousands of multiplications with R where each one does 
an unnecessary loop with strcmp?

Sarcasm? Do you have an example where you cannot get it once into a local variable and reuse?

E.g., if R is a part of MATPRODUCT, the direct pointer is already checked out.

> If you have MatSetNullSpaceMatrix than from the user standpoint it is the same as having QPSetOperatorNullSpace.

Not really. I have to include QP headers, create QP and put the matrix into the QP although I'm not interested in that :-)

Moreover, MATINV obviously needs to have access to the nullspace so it cannot be _only_ in the QP struct.

> Well, then why won't we move it out of MatInv (could be even into PETSc)?

Well, that's exactly my main point.

> I wouldn't mind if it is part of the matrix. I would mind if there was a performance hit due to it not being a part of a readily accessible struct.

> We could propose MatNullSpace having both vecs and mat, being able to set either and get one or other. The default GS for MatRemoveNullSpace() can do I-RR^T when the mat is set.

OK, I think you are very focused on the implementation detail of using PetscObjectCompose. I don't advocate that. My point is the nullspace basis is the feature of the matrix, nothing else. It would of course be better if it part of the Mat struct directly but we will need to defend it in PETSc - I'm not against.

Alternatively, we can use our own Mat struct which is exactly the same as PETSc one modulo additional entries at the end. Such can be cast to Mat. The same principle as casting to PetscObject. But it's liability to keep it in sync with PETSc.

**Comment by jkruzik** [comment]
> > Am I understanding correctly that you would do a conversion every time before a MatMult? Do you realize that such a conversion needs to copy all the vectors values into a contiguous array?
> 
> You can do that at a cost of single pointer copy in case of a dense matrix. Speaking about sparse matrices it indeed doesn't make much sense.
> 
I do not understand. To convert an array of Vecs into a dense matrix, you need to allocate a contiguous array of PetscScalar and copy the individual arrays from the Vecs into this contiguous array.
> > Fair enough. What about having a couple of thousands of multiplications with R where each one does
> > an unnecessary loop with strcmp?
> 
> Sarcasm? 

What? No, I was a bit sarcastic in the first answer after reopen but not this one ;)
>Do you have an example where you cannot get it once into a local variable and reuse?
> 
```
for ...:
	...
	Function() = {...GetNullSpaceMat(,R), MatMult(R,...)}
	...
```
also, see the first point if you want to store the null space as vecs
> E.g., if R is a part of MATPRODUCT, the direct pointer is already checked out.
> 
> > If you have MatSetNullSpaceMatrix than from the user standpoint it is the same as having QPSetOperatorNullSpace.
> 
> Not really. I have to include QP headers, create QP and put the matrix into the QP although I'm not interested in that :-)
> 
Why would you use PERMON without QP headers, QP, etc.?
> Moreover, MATINV obviously needs to have access to the nullspace so it cannot be _only_ in the QP struct.
> 
> > Well, then why won't we move it out of MatInv (could be even into PETSc)?
> 
> Well, that's exactly my main point.
> 
Great!
> > I wouldn't mind if it is part of the matrix. I would mind if there was a performance hit due to it not being a part of a readily accessible struct.
> 
> > We could propose MatNullSpace having both vecs and mat, being able to set either and get one or other. The default GS for MatRemoveNullSpace() can do I-RR^T when the mat is set.
> 
> OK, I think you are very focused on the implementation detail of using PetscObjectCompose. I don't advocate that. My point is the nullspace basis is the feature of the matrix, nothing else. It would of course be better if it part of the Mat struct directly but we will need to defend it in PETSc - I'm not against.
> 
I agree. Null space is a property of an operator. On the other hand, we use it quite a lot in QPs and we can't even solve some QPs without having a nullspace so I think having it in a QP struct is not a bad place. The question is how can we make it a matrix property so that it does not become a potential performance hit. I am not convinced a Compose() is the way to go. I would like to use MatNullSpace but from the points above I believe the nullspace needs to be kept as a matrix.

I think of Compose() as a way to glue something to an object for which you did not make a place in your object struct. Since you did not think it is important enough to have a place in your struct it should be something you need to access only occasionally (because the access through Query() is expensive). If you composed a lot of objects to something then run Query() on the last one in a long loop you would do a massive amount of strcmp().
> Alternatively, we can use our own Mat struct which is exactly the same as PETSc one modulo additional entries at the end. Such can be cast to Mat. The same principle as casting to PetscObject. But it's liability to keep it in sync with PETSc.

I do not want to duplicate Mat struct and do the syncs, especially because it is a PETSc internal struct that can be changed even on a maint branch.

**Comment by jkruzik** [comment]
I withdraw the objection to use ObjectCompose() to glue a null space to a matrix. It is not a nice way of doing things but the performance problem with many Queries() is indeed avoidable in most cases and the problem is not present in the current code.

**Comment by haplav** [comment]
Cool. I will try it in near future.

> I do not understand. To convert an array of Vecs into a dense matrix, you need to allocate a contiguous array of PetscScalar and copy the individual arrays from the Vecs into this contiguous array.

Not if the vectors are built atop the contiguous array already. This can be guaranteed if _we_ generate the vectors. Not sure how to enforce it for user-specified. MatNullSpaceCreate does not spoil it as it just maintains pointers to the vectors.

But I suggest to put off further discussion for later because PETSc changes will be needed first anyway and let's see what wider discussion with PETSc devs brings.

**Comment by jkruzik** [comment]
> Not sure how to enforce it for user-specified.
> 
We can't enforce it. But we could check if the vectors arrays are contiguous (we know the size and the pointer to the first element of the individual arrays).
> But I suggest to put off further discussion for later because PETSc changes will be needed first anyway and let's see what wider discussion with PETSc devs brings.

OK
